### PR TITLE
Faster conversion from u128/i128 to PyLong with abi3

### LINKED ISSUE #1314: Use shifting and masking to convert between 128-bit Rust integers and Python integers
@kngwyu in https://github.com/PyO3/pyo3/issues/1310#issuecomment-743722886_
> >  to use PyNumber_ APIs to shift and mask the number to get two values that fit into u64's.
> 
> I tried that first but I found it too complex, especially for `i128`, though it is faster...

We should use the faster method anyway, I don't think it'll be that much more complex.

### PR DESCRIPTION
**Pull request by kngwyu**
Fixes #1314 
I hope it's correct...

Also, adding 30 lines is acceptable for such a minor feature?

### CODE REVIEW

**Comment by kngwyu** [comment]
Now I understand why I though this approach too complex: I was thinking of using `lshift` and `add`, but `lshift` and `or` is easier.

File: src/types/num.rs [hunk 1]
```diff
@@ -261,9 +270,14 @@ mod slow_128bit_int_conversion {
...
-                    let mut bytes = [0u8; std::mem::size_of::<$rust_type>()];
+                    let mut bytes = [0u8; BYTE_SIZE];
+                    let kwargs = if $is_signed {
+                        Some([("signed", true)].into_py_dict(py))
+                    } else {
+                        None
+                    };
                     let pybytes: &PyBytes = num
-                        .call_method("to_bytes", (bytes.len(), "little"), kwargs(py, $is_signed))?
+                        .call_method("to_bytes", (bytes.len(), "little"), kwargs)?
```
Comment (birkenfeld) [inline src/types/num.rs:280]: this seems to revert a refactoring, or did you want to change the FromPy part as well?

Comment (kngwyu) [reply src/types/num.rs:280]: Since now the `kwargs` function is called once, I moved it in the function.

File: src/types/num.rs [hunk 2]
```diff
@@ -243,12 +243,21 @@ mod slow_128bit_int_conversion {
 
             impl IntoPy<PyObject> for $rust_type {
                 fn into_py(self, py: Python) -> PyObject {
-                    let bytes = PyBytes::new(py, &self.to_le_bytes());
-                    let longtype = unsafe { PyType::from_type_ptr(py, &mut ffi::PyLong_Type) };
-                    longtype
-                        .call_method("from_bytes", (bytes, "little"), kwargs(py, $is_signed))
-                        .expect("Integer conversion (u128/i128 to PyLong) failed")
-                        .into_py(py)
+                    let (first, last) = split_128int(self.to_le_bytes());
```
Comment (davidhewitt) [inline src/types/num.rs:246]: Can I suggest we use the terminology `lower` / `upper` instead of `first` / `last`? For me it's easier to understand which bits are most significant / least significant with that terminology.

```suggestion
                    let (lower, upper) = split_128int(self.to_le_bytes());
```

**Review by davidhewitt — COMMENTED** [review]
Thanks. It's 14 more lines of code total for a more efficient solution so I think this is a desirable change. I expect abi3 builds to become quite common once we have them working.

I think it'd be quite good to have a test here to be sure that we've got the right solution without any edge cases.

`proptest` can be pretty nice for tests like this where data is passed in a roundtrip. I think we can have a test like this, for example:

```rust
use proptest::prelude::*;

proptest! {
    #[test]
    fn test_i128_roundtrip(x: i128) {
        Python::with_gil(|py| {
            let x_py = x.into_py(py);
            py_run!(py, x_py, format!("assert x_py == {}", x));
            let roundtripped: i128 = x_py.extract().unwrap();
            assert_eq!(x, roundtripped);
        })
    }
}
```

and similar for u128.

File: src/types/num.rs (hunk 2 above)
Comment (kngwyu) [reply src/types/num.rs]: Thanks, I wanted better names too 😉

**Comment by kngwyu** [comment]
Just a status update: renamed some variables and re-implement PyLong to i128/u128 conversion using `PyLong_AsUnsignedLongLongMask` and shift.
Proptest looks good but I'm still reading the document.

File: src/types/num.rs [hunk 3]
```diff
@@ -243,37 +241,49 @@ mod slow_128bit_int_conversion {
...
 
             impl<'source> FromPyObject<'source> for $rust_type {
                 fn extract(ob: &'source PyAny) -> PyResult<$rust_type> {
                     let py = ob.py();
-                    let num: &PyLong = unsafe {
-                        crate::FromPyPointer::from_owned_ptr_or_err(
+                    unsafe {
+                        let lower = err_if_invalid_value(
                             py,
-                            ffi::PyNumber_Index(ob.as_ptr()),
```
Comment (davidhewitt) [inline src/types/num.rs:261]: I think the call to `PyNumber_Index` is still needed before calling `PyLong_AsUnsignedLongLongMask`, so that types which implement `__index__` are converted to `PyLong`?

**Review by davidhewitt — COMMENTED** [review]
Nice! Couple more suggestions...

File: src/types/num.rs (hunk 3 above)
Comment (kngwyu) [reply src/types/num.rs:261]: No, `PyLong_AsUnsigned...` calls it internally.

File: src/types/num.rs [hunk 4]
```diff
@@ -243,37 +241,49 @@ mod slow_128bit_int_conversion {
...
-                    let bytes = PyBytes::new(py, &self.to_le_bytes());
-                    let longtype = unsafe { PyType::from_type_ptr(py, &mut ffi::PyLong_Type) };
-                    longtype
-                        .call_method("from_bytes", (bytes, "little"), kwargs(py, $is_signed))
-                        .expect("Integer conversion (u128/i128 to PyLong) failed")
-                        .into_py(py)
+                    let le_bytes = self.to_le_bytes();
+                    let lower = u64::from_le_bytes(slice_to_bytearr(&le_bytes[..BYTE_SIZE / 2]));
+                    let upper =
+                        <$half_type>::from_le_bytes(slice_to_bytearr(&le_bytes[BYTE_SIZE / 2..]));
```
[the remark this answers is not in the document]
Comment (kngwyu) [reply src/types/num.rs:247]: Looks nice 👍🏽

**Comment by kngwyu** [comment]
Added an overflow test for i128.
And for `proptest`, I think in this case `max/min` is sufficiently strong and we don't *strongly* need randomized tests.

**Comment by birkenfeld** [comment]
There should also be a test with values in the `i/u64` range, to ensure that the two halves are not mixed up.

File: src/types/num.rs (hunk 3 above)
Comment (davidhewitt) [reply src/types/num.rs:261]: Ah perfect, thanks for checking!

**Comment by davidhewitt** [comment]
Agreed; if we're not proptesting we should make sure the tests we do have cover for mistakes like halves being mixed up. We know it's correct now but who knows what will happen next time I try to refactor this code 👀

**Comment by programmerjake** [comment]
Why are you using all the le bytes methods when you can just use:
```rust
let v: u128 = ...;
let low_half = v as u64;
let high_half = (v >> 64) as u64;
```
and
```rust
let v: i128 = ...;
let low_half = v as u64;
let high_half = (v >> 64) as i64;
```
and then on the python side:
```Python
(high_half << 64) | low_half
```

**Comment by programmerjake** [comment]
Good test values are:
```rust
let u128_test_value = 0xFF0102030405060788090A0B0C0D0E0Fu128;
let i128_test_value = u128_test_value as i128;
```
since it tests the halves are in the right order with the right signedness.

**Comment by davidhewitt** [comment]
> Why are you using all the le bytes methods when you can just use:

😅 great suggestion, thanks! I think when this implementation started with `to_bytes` we got stuck on one idea too much!

**Comment by kngwyu** [comment]
> Why are you using all the le bytes methods when you can just use:

Wow, thanks. It seems like I wrote too many TeX documents to forget some bit hacks 🙄

**Comment by kngwyu** [comment]
Added a basic proptest but is this sufficient?

File: src/types/num.rs [hunk 5]
```diff
@@ -243,43 +234,79 @@ mod slow_128bit_int_conversion {
...
-                        )
-                    }?;
-                    let mut bytes = [0u8; std::mem::size_of::<$rust_type>()];
-                    let pybytes: &PyBytes = num
-                        .call_method("to_bytes", (bytes.len(), "little"), kwargs(py, $is_signed))?
-                        .extract()?;
-                    bytes.copy_from_slice(pybytes.as_bytes());
-                    Ok(<$rust_type>::from_le_bytes(bytes))
+                            -1 as _,
+                            ffi::PyLong_AsUnsignedLongLongMask(ob.as_ptr()),
```
Comment (programmerjake) [inline src/types/num.rs:262]: I think you'd probably want an assertion of some sort that `PyLong_UnsignedLongLongMask` returns the expected type, since C/C++ don't guarantee that it's `u64` -- it could be `u128` or some other type on an unusual OS.

**Comment by programmerjake** [comment]
You accidentally changed `src/lib.rs` to an executable file.

**Comment by programmerjake** [comment]
Btw, sorry to sound all negative, thanks for all your work!

**Comment by programmerjake** [comment]
other than all that, looks good to me!

File: src/types/num.rs (hunk 5 above)
Comment (kngwyu) [reply src/types/num.rs:262]: Hmm 🤔 , but even the Rust stdlib does not support it https://doc.rust-lang.org/nightly/std/os/raw/type.c_ulonglong.html.

**Comment by kngwyu** [comment]
> Btw, sorry to sound all negative, thanks for all your work!

No worry, thank you for your comments 👍🏼

File: src/types/num.rs (hunk 5 above)
Comment (programmerjake) [reply src/types/num.rs:262]: The docs are just showing the type used on the system they used to build the docs (`x86_64-unknown-linux-gnu` iirc), it doesn't mean it will always be `u64`.

Comment (programmerjake) [reply src/types/num.rs:262]: For an example, see https://doc.rust-lang.org/nightly/std/os/raw/type.c_long.html

Comment (kngwyu) [reply src/types/num.rs:262]: I meant the description: 
> The C standard technically only requires that this type be an unsigned integer with the size of a long long, although in practice, no system would have a long long that is not a u64, as most systems do not have a standardised u128 type.

Also, it has no `#[cfg` block:
image
So I think we don't need to consider the system where ull is u128 *for now*

Comment (programmerjake) [reply src/types/num.rs:262]: ok, yeah. I did check both the AS/400 and RISC-V 128-bit compilers (or at least their ABI proposals/docs) and they both have a 64-bit ull -- AS/400 just doesn't have a 128-bit int type (no intptr_t at all), and RV128 proposes `long long long` for 128-bit types.

**Review by davidhewitt — APPROVED** [review]
Looks good! Just need to `chmod -x src/lib.rs` before merging. Not sure how that happened 🤔

**Comment by kngwyu** [comment]
Thanks!
> Not sure how that happened

Maybe my emacs setting does something wrong, but I'm also not sure 😓
